### PR TITLE
Split Init.hs into smaller, more focused modules.

### PR DESCRIPTION
**Pull request by m-renaud**
This splits the monolithic `Init.hs` module which runs the `cabal init` command into smaller sub-modules. This PR adds 5 new module:

- `Init.Command` - Main `cabal init` logic that ties everything together.
- `Init.Defaults` - Default values for items not specified in config or arguments
- `Init.FileCreators` - All the `createXyz` functions which write files (LICENSE, Main.hs, etc.)
- `Init.Prompt` - Utilities for prompting the user for selections
- `Init.Shared` - Misc utilities shared across the above modules

I used the Git's octopus merge strategy which should maintain blame information from the original `Init.hs` source code. Here's an excerpt from `git blame`:

```
$ git blame Distribution/Client/Init/Command.hs
375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 109) initCabal :: Verbosity
1375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 110)           -> PackageDBStack
ba5c55c442 cabal-install/Distribution/Client/Init.hs         (Edsko de Vries         2016-01-04 16:39:54 +0100 111)           -> RepoContext
1375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 112)           -> Compiler
be5eadab33 cabal-install/Distribution/Client/Init.hs         (Mikhail Glushenkov     2016-08-16 15:38:14 +0200 113)           -> ProgramDb
1375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 114)           -> InitFlags
1375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 115)           -> IO ()
d8ef88a96d cabal-install/Distribution/Client/Init.hs         (Mikhail Glushenkov     2016-09-06 03:44:06 +0100 116) initCabal verbosity packageDBs repoCtxt comp progdb initFlags = do
1375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 117) 
d8ef88a96d cabal-install/Distribution/Client/Init.hs         (Mikhail Glushenkov     2016-09-06 03:44:06 +0100 118)   installedPkgIndex <- getInstalledPackages verbosity comp packageDBs progdb
ba5c55c442 cabal-install/Distribution/Client/Init.hs         (Edsko de Vries         2016-01-04 16:39:54 +0100 119)   sourcePkgDb <- getSourcePackages verbosity repoCtxt
1375d0651d cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2011-10-25 05:09:00 +0000 120) 
4081112cf5 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2009-10-11 16:56:44 +0000 121)   hSetBuffering stdout NoBuffering
4081112cf5 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2009-10-11 16:56:44 +0000 122) 
7fdbabda05 cabal-install/Distribution/Client/Init.hs         (Fabián Orccón          2015-02-21 09:41:31 -0500 123)   initFlags' <- extendFlags installedPkgIndex sourcePkgDb initFlags
4081112cf5 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2009-10-11 16:56:44 +0000 124) 
67ed3a6231 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2015-07-20 14:18:40 -0500 125)   case license initFlags' of
67ed3a6231 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2015-07-20 14:18:40 -0500 126)     Flag PublicDomain -> return ()
67ed3a6231 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2015-07-20 14:18:40 -0500 127)     _                 -> writeLicense initFlags'
4d6175ccb3 cabal-install/Distribution/Client/Init.hs         (Toshio Ito             2015-03-04 08:35:13 +0900 128)   writeChangeLog initFlags'
0723d6f11c cabal-install/Distribution/Client/Init.hs         (Matt Renaud            2018-12-02 22:26:36 -0800 129)   createDirectories (sourceDirs initFlags')
d59a224dc3 cabal-install/Distribution/Client/Init.hs         (Matt Renaud            2019-02-07 13:23:50 -0800 130)   createLibHs initFlags'
d59a224dc3 cabal-install/Distribution/Client/Init.hs         (Matt Renaud            2019-02-07 13:23:50 -0800 131)   createDirectories (applicationDirs initFlags')
ad95fb2418 cabal-install/Distribution/Client/Init.hs         (Dennis Gosnell         2015-03-19 00:35:47 +0900 132)   createMainHs initFlags'
e3475bb967 cabal-install/Distribution/Client/Init.hs         (Mikhail Glushenkov     2019-01-25 02:00:51 +0000 133)   -- If a test suite was requested and this is not an executable-only
e3475bb967 cabal-install/Distribution/Client/Init.hs         (Mikhail Glushenkov     2019-01-25 02:00:51 +0000 134)   -- package, then create the "test" directory.
8fa08ce42c cabal-install/Distribution/Client/Init/Command.hs (Matt Renaud            2020-04-01 15:28:47 -0700 135)   createTestSuiteIfEligible initFlags'
4081112cf5 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2009-10-11 16:56:44 +0000 136)   success <- writeCabalFile initFlags'
4081112cf5 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2009-10-11 16:56:44 +0000 137) 
4081112cf5 cabal-install/Distribution/Client/Init.hs         (Brent Yorgey           2009-10-11 16:56:44 +0000 138)   when success $ generateWarnings initFlags'

```

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: cabal-install/Distribution/Client/Init/Shared.hs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+--
+-- Maintainer  :  cabal-devel@haskell.org
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- Shared utilities used by multiple cabal init modules.
+--
+-----------------------------------------------------------------------------
+
+module Distribution.Client.Init.Shared (
```
Comment (phadej) [inline cabal-install/Distribution/Client/Init/Shared.hs:15]: We name such modules `Utils`

Comment (m-renaud) [reply cabal-install/Distribution/Client/Init/Shared.hs:15]: Done.

**Comment by m-renaud** [comment]
FYI if I squash we lose blame history in the forked files, are you alright with that?

**Comment by phadej** [comment]
Yes. Having 16 commits for a conceptually one operation is not good. I do use `git bisect`.

**Comment by m-renaud** [comment]
SG, will squash to one commit and merge.

**Comment by m-renaud** [comment]
Whoa, something went wrong locally, need to fix this up.

**Pull request closed without merging**